### ISSUE TITLE
Minimum staked percentage of total supply for the DAO to be active

### ISSUE THREAD

**Issue by JakeHartnell** [DA0-DA0/dao-contracts#160]
@ezekiiel, think this would solve the race issue.

**Comment by ezekiiel** [comment]
This sounds like a nice idea. Getting the appropriate percentage requirement right might take some experimentation.

What do you think about allowing the dao to be created with an existing staking contract? Other contracts can stake on behalf of a wallet so this would allow the issuance of staked tokens before the dao was created.

**Comment by ezekiiel** [comment]
Some thoughts:

We likely want to provide both an absolute count and percentage threshold for staked tokens. Something like:

```rust
pub enum StakingThreshold {
  AbsoluteCount { count: Uint128 },
  Percentage { percent: Decimal },
}
```

IMO this may be something to implement in a proposal module. The trouble with implementing this in a voting module is that we still want queries about voting power to work if that threshold isn't met. People need to know how much they have staked. As the query is the only interface between voting modules and proposal modules I suspect implementing this in a voting module would be hard.

Instead, it seems reasonable that a governance module supports this. There is already a `TokenContract {}` query which voting modules that have an associated token should support. If we just didn't allow creation of proposals if the threshold wasn't met this would likely solve our issues.

No need to check on vote if we do that either as voting power will be determined at a height where the threshold is met.

**Comment by Callum-A** [comment]
> Some thoughts:
> 
> We likely want to provide both an absolute count and percentage threshold for staked tokens. Something like:
> 
> ```rust
> pub enum StakingThreshold {
>   AbsoluteCount { count: Uint128 },
>   Percentage { percent: Decimal },
> }
> ```
> 
> IMO this may be something to implement in a proposal module. The trouble with implementing this in a voting module is that we still want queries about voting power to work if that threshold isn't met. People need to know how much they have staked. As the query is the only interface between voting modules and proposal modules I suspect implementing this in a voting module would be hard.
> 
> Instead, it seems reasonable that a governance module supports this. There is already a `TokenContract {}` query which voting modules that have an associated token should support. If we just didn't allow creation of proposals if the threshold wasn't met this would likely solve our issues.
> 
> No need to check on vote if we do that either as voting power will be determined at a height where the threshold is met.

Thinking of getting to this soon, just wondering how on the proposal side would we determine if Staking is being used to derive voting power as surely we cannot assume every DAO is using staked balances to derive the voting power? Would we need a flag to determine if we use staking or not to enable this?

**Comment by Callum-A** [comment]
What are we thinking on this? Should we assume the voting module has a `TokenContract {}` endpoint and fail safely if not and use that to then determine if the threshold is met? Or is there a better way

**Comment by ezekiiel** [comment]
This is a really good question. I think it would be reasonable in these situations to assume that the voting power module is using staking. If a user tries to set this up without using staking the UI can warn really loudly that it won't work.

For this reason though this should probably be a separate voting module that does this than the existing one as the existing one needs to support multisigs. This would be a really great chance to move more into the `voting` package.

**Comment by Callum-A** [comment]
> This is a really good question. I think it would be reasonable in these situations to assume that the voting power module is using staking. If a user tries to set this up without using staking the UI can warn really loudly that it won't work.
> 
> 
> 
> For this reason though this should probably be a separate voting module that does this than the existing one as the existing one needs to support multisigs. This would be a really great chance to move more into the `voting` package.

Probably miss understanding but how would this be implemented on the voting module side? As these are only used to query for power?

**Comment by ezekiiel** [comment]
Currently voting power modules with tokens should implement a `TokenContract {}` query. A governance module that wants to enforce a minimum staked balance can assume that staking is being used and look at `total_voting_power / token_supply` to determine if the staking threshold has been met. It can get the token supply via a second query to the token contract.

**Comment by Callum-A** [comment]
> Currently voting power modules with tokens should implement a `TokenContract {}` query. A governance module that wants to enforce a minimum staked balance can assume that staking is being used and look at `total_voting_power / token_supply` to determine if the staking threshold has been met. It can get the token supply via a second query to the token contract. Am I understanding the question correctly?

Yeah I just am worried at users bricking their DAO but if this assumption is fine and we can flag it on the FE, I can get this done.
